health-checker/src/main/java/io/stargate/health/HealthCheckerActivator.java

```diff
@@ -19,8 +15,13 @@ public void start(BundleContext context) {
     this.context = context;
     log.info("Starting healthchecker....");
 
-    Metrics metrics = new MetricsImpl();
-    context.registerService(Metrics.class, metrics, null);
+    ServiceReference<?> metricsReference = context.getServiceReference(Metrics.class.getName());
+
+    if (metricsReference == null) {
```

Comment:
is there a possibility that the metrics module cannot be loaded but this is expected? For example, if the client doesn't want to include metrics on their classpath? Maybe in this situation, we shouldn't fail fast?

Reply:
Metrics gathering is always enabled.
Health checks depends on metrics being gathered.

Reply:
I'm not sure OSGi guarantees any particular order of bundle activation. So even though we start the metrics service unconditionally, it may not be available yet when this method runs.... I think :thinking:

Reply:
I believe I saw something like a helper lib to manage inter-service dependencies on the web, but I lost the tab for it :)

Reply:
oh, interesting...

I was thinking OSGi used the import packages as a way to reorder activation. We should find a way then...

Reply:
something like this perhaps? https://www.osgi.org/wp-content/uploads/AutoServDependencyMgmt_byHall_Cervantes.pdf

Reply:
also: https://felix.apache.org/documentation/subprojects/apache-felix-dependency-manager/guides/bundles-and-dependencies.html

Reply:
AFAIK, imports are just for managing class loading :thinking: but I'm not an OSGi expert :)

Reply:
oh my, no easy way to set (ordered) dependencies between service instances, I'm hating OSGi more each day... 

I think the shorter path for now would be to use `serviceChanged()`... I'll adapt the PR.

Reply:
Check out `ServiceTracker`, cf. https://github.com/stargate/api-extensions/blob/select-by-pk/sql/server/src/main/java/io/stargate/api/sql/server/SqlActivator.java#L47

Reply:
Bundle activation will occur as dependencies are resolved. So persistence for example will activate first since it has no dependencies but something like graphql will activate later because it waits on persistence -> auth -> graphql.

Reply:
Thanks for the clarification @dougwettlaufer !